libcontainer/configs/config.go

```diff
@@ -159,4 +165,75 @@ type Config struct {
 	// A number of rules are given, each having an action to be taken if a syscall matches it.
 	// A default action to be taken if no rules match is also given.
 	Seccomp *Seccomp `json:"seccomp"`
+
+	// Hooks are a collection of actions to perform at various container lifecycle events.
+	// Hooks are not able to be marshaled to json but they are also not needed to.
+	Hooks *Hooks `json:"-"`
+}
+
+type Hooks struct {
+	// Prestart commands are executed after the container namespaces are created,
+	// but before the user supplied command is executed from init.
+	Prestart []Hook
+
+	// PostStop commands are executed after the container init process exits.
+	Poststop []Hook
+}
+
+// HookState is the payload provided to a hook on execution.
+type HookState struct {
+	ID  string `json:"id"`
+	Pid int    `json:"pid"`
+}
+
+type Hook interface {
+	// Run executes the hook with the provided state.
+	Run(*HookState) error
```

Comment:
`HookState` very simple "immutable" struct. Can be passed everywhere by value to avoid confusion.